File: python/safe_wallet/6_market_buy_order.py
"""
6_market_buy_order.py — FOK market BUY at best ask (Safe pUSD) → poll until filled.

Safe-based: pUSD collateral locked in the Safe. signatureType=2.
Run 7_market_sell_order.py next to sell the YES tokens back.

Run:
    python 6_market_buy_order.py

Requires: PRIVATE_KEY, SAFE_ADDRESS, CLOB_V2_BASE_URL, POLYGON_RPC_URL in .env
"""

import os
import time
import requests
from dotenv import load_dotenv
from web3 import Web3
from py_clob_client_v2.client import ClobClient
from py_clob_client_v2.clob_types import MarketOrderArgsV2, PartialCreateOrderOptions, OrderType, BalanceAllowanceParams, AssetType

load_dotenv(dotenv_path=os.path.join(os.path.dirname(__file__), ".env"))

PRIVATE_KEY  = os.environ["PRIVATE_KEY"]
SAFE_ADDRESS = os.environ["SAFE_ADDRESS"]
CLOB_HOST    = os.environ.get("CLOB_V2_BASE_URL", "https://clob-v2.polymarket.com")
RPC_URL      = os.environ.get("POLYGON_RPC_URL", "https://polygon-bor-rpc.publicnode.com")
EVENT_ID     = "73106"
PUSD         = Web3.to_checksum_address("0xC011a7E12a19f7B1f670d46F03B03f3342E82DFB")
BUY_SPEND    = 2.0   # dollars

ERC20_ABI = [{"name": "balanceOf", "type": "function", "stateMutability": "view",
              "inputs": [{"name": "account", "type": "address"}],
              "outputs": [{"name": "", "type": "uint256"}]}]

w3   = Web3(Web3.HTTPProvider(RPC_URL))
SAFE = Web3.to_checksum_address(SAFE_ADDRESS)
pusd_contract = w3.eth.contract(address=PUSD, abi=ERC20_ABI)

def get_safe_pusd() -> float:
    return pusd_contract.functions.balanceOf(SAFE).call() / 1e6

# ── Init client ───────────────────────────────────────────────────────────────
temp_client = ClobClient(CLOB_HOST, chain_id=137, key=PRIVATE_KEY, signature_type=2, funder=SAFE_ADDRESS)
creds       = temp_client.create_or_derive_api_key()
client      = ClobClient(CLOB_HOST, chain_id=137, key=PRIVATE_KEY, creds=creds, signature_type=2, funder=SAFE_ADDRESS)
print(f"Safe    : {SAFE}")
print(f"API key : {creds.api_key} ✓\n")

# ── Sync on-chain balances/allowances to CLOB server ─────────────────────────
client.update_balance_allowance(BalanceAllowanceParams(asset_type=AssetType.COLLATERAL))
print("Balance/allowance synced ✓\n")

# ── Market info ───────────────────────────────────────────────────────────────
event    = requests.get(f"https://gamma-api.polymarket.com/events/{EVENT_ID}").json()
cond_id  = event["markets"][0]["conditionId"]
info     = client.get_clob_market_info(cond_id)
tokens   = info.get("t", [])
token_id = next((t["t"] for t in tokens if t["o"] == "Yes"), "")
tick_size = str(info.get("mts", "0.01"))
neg_risk  = info.get("nr", False)
options  = PartialCreateOrderOptions(tick_size=tick_size, neg_risk=neg_risk)

print(f"Market: tick={tick_size}  negRisk={neg_risk}  token={token_id[:16]}…")

# ── Order book — best ask ─────────────────────────────────────────────────────
book = client.get_order_book(token_id)
asks = sorted(book["asks"], key=lambda a: float(a["price"]))
if not asks:
    print("No asks — can't market buy"); exit(1)

best_ask   = float(asks[0]["price"])
bal_before = get_safe_pusd()

print(f"Best ask  : ${best_ask}  pUSD (Safe): ${bal_before:.4f}")
print(f"Buy spend : ${BUY_SPEND}\n")

if bal_before < BUY_SPEND + 0.5:
    print("⚠ Not enough pUSD in Safe."); exit(1)

# ── FOK market BUY (amount = dollars to spend) ───────────────────────────────
buy_resp = client.create_and_post_market_order(
    MarketOrderArgsV2(token_id=token_id, amount=BUY_SPEND, side="BUY", price=best_ask),
    options, OrderType.FOK,
)
buy_id = buy_resp.get("orderID") if isinstance(buy_resp, dict) else buy_resp.orderID
status = buy_resp.get("status")  if isinstance(buy_resp, dict) else buy_resp.status
print(f"Buy order : {buy_id}  status: {status}")
if not buy_id:
    print(f"Rejected  : {buy_resp.get('errorMsg') if isinstance(buy_resp, dict) else buy_resp}"); exit(1)

# ── Poll until filled (up to 30 s) ────────────────────────────────────────────
print("Polling for fill…")
for i in range(30):
    time.sleep(1)
    open_orders = client.get_open_orders() or []
    ids = [o.get("id", o.get("orderID", "")) for o in open_orders]
    if buy_id not in ids:
        print("✓ Filled")
        break
    print(".", end="", flush=True)
else:
    print("\n⚠ Still open after 30 s")

File: typescript/safe_wallet/6_market_buy_order.ts
/**
 * 6_market_order.ts — Market BUY (FOK) → poll fill → market SELL back.
 *
 * Safe wallet: signatureType=2, funderAddress=SAFE. pUSD held by Safe.
 *
 * Run:
 *   npm run 6:market
 */

import "dotenv/config";
import { createWalletClient, createPublicClient, http, erc20Abi, formatUnits } from "viem";
import { polygon } from "viem/chains";
import { privateKeyToAccount } from "viem/accounts";
import { ClobClient, Side, OrderType, type UserMarketOrderV2 } from "@polymarket/clob-client-v2";

const PRIVATE_KEY  = process.env.PRIVATE_KEY!;
const SAFE_ADDRESS = process.env.SAFE_ADDRESS! as `0x${string}`;
const CLOB_HOST    = process.env.CLOB_V2_BASE_URL ?? "https://clob-v2.polymarket.com";
const RPC_URL      = process.env.POLYGON_RPC_URL  ?? "https://polygon-bor-rpc.publicnode.com";
const EVENT_ID     = "73106";
const PUSD         = "0xC011a7E12a19f7B1f670d46F03B03f3342E82DFB" as `0x${string}`;
const MIN_SIZE     = 5;
const BUY_SPEND    = 2;

const account      = privateKeyToAccount(`0x${PRIVATE_KEY}` as `0x${string}`);
const walletClient = createWalletClient({ account, chain: polygon, transport: http(RPC_URL) });
const publicClient = createPublicClient({ chain: polygon, transport: http(RPC_URL) });

console.log("EOA  :", account.address);
console.log("Safe :", SAFE_ADDRESS);
console.log("CLOB :", CLOB_HOST, "\n");

function sleep(ms: number) { return new Promise(r => setTimeout(r, ms)); }

async function getPusd(): Promise<number> {
  const raw = await publicClient.readContract({ address: PUSD, abi: erc20Abi, functionName: "balanceOf", args: [SAFE_ADDRESS] });
  return parseFloat(formatUnits(raw, 6));
}

// ── Init ClobClient V2 (Safe) ─────────────────────────────────────────────────

const tempClient = new ClobClient({
  host: CLOB_HOST, chain: 137, signer: walletClient as never,
  signatureType: 2, funderAddress: SAFE_ADDRESS,
});
const creds = await (tempClient as any).createOrDeriveApiKey();
const client = new ClobClient({
  host: CLOB_HOST, chain: 137, signer: walletClient as never,
  signatureType: 2, funderAddress: SAFE_ADDRESS,
  creds: { key: creds.key, secret: creds.secret, passphrase: creds.passphrase },
});
console.log("API key :", creds.key, "✓");

// ── Market info ───────────────────────────────────────────────────────────────

const event    = await fetch(`https://gamma-api.polymarket.com/events/${EVENT_ID}`).then(r => r.json()) as any;
const condId   = event.markets?.[0]?.conditionId as string;
const info     = await (client as any).getClobMarketInfo(condId);
const tokens   = (info?.t ?? []) as { t: string; o: string }[];
const tokenId  = tokens.find(t => t.o === "Yes")?.t ?? "";
const tickSize = (info?.mts?.toString() ?? "0.01") as "0.1" | "0.01" | "0.001" | "0.0001";
const negRisk  = info?.nr ?? false;

// ── Book snapshot ─────────────────────────────────────────────────────────────

const book    = await (client as any).getOrderBook(tokenId);
const asks    = [...(book?.asks ?? [])].sort((a: any, b: any) => parseFloat(a.price) - parseFloat(b.price));
if (asks.length === 0) { console.error("No asks — can't market buy"); process.exit(1); }
const bestAsk = parseFloat(asks[0].price);

// ── pUSD balance (from Safe) ──────────────────────────────────────────────────

const balBefore = await getPusd();
console.log(`pUSD (Safe): $${balBefore.toFixed(6)}`);
if (balBefore < BUY_SPEND + 0.5) { console.error("⚠ Not enough pUSD in Safe."); process.exit(1); }

console.log(`Best ask: $${bestAsk}  Spending: $${BUY_SPEND}`);

// ── FOK market BUY ────────────────────────────────────────────────────────────

console.log(`\nPlacing FOK BUY: $${BUY_SPEND} @ $${bestAsk}…`);
const buyOrder: UserMarketOrderV2 = { tokenID: tokenId, price: bestAsk, amount: BUY_SPEND, side: Side.BUY };
const buyResp = await client.createAndPostMarketOrder(buyOrder, { tickSize, negRisk }, OrderType.FOK);
console.log(`  Order ID : ${buyResp.orderID}  status: ${buyResp.status}`);
if (!buyResp.orderID) { console.error("Buy rejected:", (buyResp as any).errorMsg); process.exit(1); }

// ── Poll until filled ─────────────────────────────────────────────────────────

console.log("\nPolling for fill (up to 30 s)…");
let filled = false;
for (let i = 0; i < 30; i++) {
  await sleep(1000);
  const open: any[] = await (client as any).getOpenOrders() ?? [];
  if (!open.find((o: any) => o.id === buyResp.orderID)) { filled = true; break; }
  process.stdout.write(".");
}
console.log(filled ? "\n✓ Filled" : "\n⚠ Still open after 30 s");

const balAfterBuy = await getPusd();
const spent       = balBefore - balAfterBuy;
console.log(`pUSD spent: $${spent.toFixed(6)}`);
if (spent < 0.01) { console.log("⚠ FOK may not have filled (thin book). Exiting."); process.exit(0); }

// ── FOK market SELL back ──────────────────────────────────────────────────────

const bookNow  = await (client as any).getOrderBook(tokenId);
const bidsNow  = [...(bookNow?.bids ?? [])].sort((a: any, b: any) => parseFloat(b.price) - parseFloat(a.price));
if (bidsNow.length === 0) { console.log("⚠ No bids — skipping sell."); process.exit(0); }
const bestBid  = parseFloat(bidsNow[0].price);
const sellSize = Math.max(MIN_SIZE, Math.floor((spent / bestAsk) * 100) / 100);

console.log(`\nPlacing FOK SELL: ${sellSize} shares @ $${bestBid}…`);
const sellOrder: UserMarketOrderV2 = { tokenID: tokenId, price: bestBid, amount: sellSize, side: Side.SELL };
const sellResp = await client.createAndPostMarketOrder(sellOrder, { tickSize, negRisk }, OrderType.FOK);
console.log(`  Order ID : ${sellResp.orderID}  status: ${sellResp.status}`);

// ── Final balances ────────────────────────────────────────────────────────────

let sellFilled = false;
for (let i = 0; i < 30; i++) {
  await sleep(1000);
  const open: any[] = await (client as any).getOpenOrders() ?? [];
  if (!open.find((o: any) => o.id === sellResp.orderID)) { sellFilled = true; break; }
  process.stdout.write(".");
}
console.log(sellFilled ? "\n✓ Sell filled" : "\n⚠ Sell still open");

const balFinal = await getPusd();
console.log(`\npUSD start  : $${balBefore.toFixed(6)}`);
console.log(`pUSD final  : $${balFinal.toFixed(6)}`);
console.log(`Net PnL     : $${(balFinal - balBefore).toFixed(6)}`);
